zookeeper-server/src/main/java/org/apache/zookeeper/server/SnapshotComparer.java

```diff
@@ -0,0 +1,461 @@
+import org.apache.commons.cli.OptionBuilder;
+import org.apache.commons.cli.ParseException;
+import org.apache.jute.BinaryInputArchive;
+import org.apache.jute.InputArchive;
+import org.apache.zookeeper.server.persistence.FileSnap;
+
+/**
+ * Compare two data tree snapshots, output information about the delta.
+ * Only outputs information about permanent nodes, ignoring both sessions
+ * and ephemeral nodes.
+ */
+public class SnapshotComparer {
+  private final Options options;
+  private static final String leftOption = "left";
+  private static final String rightOption = "right";
+  private static final String byteThresholdOption = "bytes";
+  private static final String nodeThresholdOption = "nodes";
+  private static final String debugOption = "debug";
+  private static final String interactiveOption = "interactive";
+
+  @SuppressWarnings("static")
+  private SnapshotComparer() {
+    options = new Options();
+    options.addOption(
+        OptionBuilder
+            .hasArg()
+            .isRequired(true)
+            .withLongOpt(leftOption)
+            .withDescription("the left snapshot file")
+            .withArgName("LEFT")
+            .withType(File.class)
+            .create("l"));
+    options.addOption(
+        OptionBuilder
+            .hasArg()
+            .isRequired(true)
+            .withLongOpt(rightOption)
+            .withDescription("the right snapshot file")
+            .withArgName("RIGHT")
+            .withType(File.class)
+            .create("r"));
+    options.addOption(
+        OptionBuilder
+            .hasArg()
+            .isRequired(true)
+            .withLongOpt(byteThresholdOption)
+            .withDescription("the node data delta size threshold, in bytes, for printing the node")
+            .withArgName("BYTETHRESHOLD")
+            .withType(File.class)
+            .create("b"));
+    options.addOption(
+        OptionBuilder
+            .hasArg()
+            .isRequired(true)
+            .withLongOpt(nodeThresholdOption)
+            .withDescription("the descendant node delta size threshold, in nodes, for printing the node")
+            .withArgName("NODETHRESHOLD")
+            .withType(String.class)
+            .create("n"));
+    options.addOption(
+        OptionBuilder
+            .hasArg()
+            .withLongOpt(debugOption)
+            .withDescription("use debug output")
+            .withArgName("DEBUG")
+            .withType(String.class)
+            .create("d"));
+    options.addOption(
+        OptionBuilder
+            .hasArg()
+            .withLongOpt(interactiveOption)
+            .withDescription("interactive mode")
+            .withArgName("INTERACTIVE")
+            .withType(String.class)
+            .create("i"));
+  }
+
+  private void usage() {
+    HelpFormatter help = new HelpFormatter();
+
+    help.printHelp(
+        120,
+        "java -cp <classPath> " + SnapshotFormatter.class.getName(),
+        "",
+        options,
```

Comment:
- SnapshotFormatter? I also notice there're some useless codes in this class.
- @hanm  Could you plz encapsulate `SnapshotComparer.java` to a scriprt just like `zkTxnLogToolkit.sh` which's more user-fridenly. and also doc the usage about this useful diff tool about snapshot in the `zookeeperTools.md`

Reply:
good catch, obviously we copied this from the SnapshotFormatter to get started. I'll clean those up, and put some documentation up. I'll probably let someone else take the work on making this more user friendly though.

Reply:
created https://issues.apache.org/jira/browse/ZOOKEEPER-3505 to follow up the shell script tool. :)

Reply:
-@mayawang   Thanks for this.ZOOKEEPER-3344 may be a good example.
-@hanm Can we move on? I remembered I had done a quick try for this tool and have no idea about how to use it :)

Reply:
@maoling Thanks! I'll work with @hanm to continue improving it.